### PR TITLE
[TORCH] Unwrap and wrap torch.return_type before and after posthooks

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* Post hooks inputs are being unwrapped from `torch.return_type` types to `torch.tensor` type.
* Pre hooks outputs are being wrapped from  `torch.tensor` to  `torch.return_type` in case torch input were wrapped in the first place.

### Reason for changes

To enable post hook insertion after torch operations which return `torch.tensor_type` values instead of `torch.tensor` values.
 
### Tests

* tests/torch/test_nncf_network.py is updated
* tests/torch/test_return_types.py is introduced

Verbally approved by @vshampor

### CODE REVIEW

File: nncf/torch/dynamic_graph/wrappers.py [hunk 1]
```diff
@@ -188,8 +190,10 @@ def _execute_op(
         if is_debug() and node is not None:
             ctx.register_node_call(node)
 
-    result = trace_tensors(result, node, ctx)
-    result = ctx.execute_post_hooks(op_address, result)
+    unwrapped_result = maybe_unwrap_from_torch_return_type(result)
+    unwrapped_result = trace_tensors(unwrapped_result, node, ctx)
+    unwrapped_result = ctx.execute_post_hooks(op_address, unwrapped_result)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/dynamic_graph/wrappers.py:195]: As return type appears only after operations like min, max, sort, and as we know no operations in torch that take return type as an input, together with @vshampor we decided to wrap only post_hooks

File: nncf/torch/dynamic_graph/wrappers.py [hunk 2]
```diff
@@ -188,8 +190,10 @@ def _execute_op(
         if is_debug() and node is not None:
             ctx.register_node_call(node)
 
-    result = trace_tensors(result, node, ctx)
-    result = ctx.execute_post_hooks(op_address, result)
+    unwrapped_result = maybe_unwrap_from_torch_return_type(result)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/dynamic_graph/wrappers.py:193]: It is necessary for training in case a post hook after a max/min/sort and etc. is present in the model. The cost of the wrapping for our target scope is two `isinstance` calls for each operation in graph as there was no post hook after a max/min/sort and etc op yet

**Comment by daniil-lyakhov** [comment]
> Some general comments:
> 
> * Could you check that operations which returns torch.return_type correctly traced?
> * Could you check that `pre_hooks` also support torch.return_type?
> * Did you consider the approach with calling `maybe_unwrap_from_torch_return_type` before hook?

* Please see `test_torch_return_types_unwrapped_for_post_hook`
* Please see https://github.com/openvinotoolkit/nncf/pull/2290#discussion_r1405919297

**Review by alexsu52 — APPROVED** [review]
LGTM

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: tests/torch/test_return_types.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# limitations under the License.
+import torch
+
+from nncf.torch.return_types import _TORCH_RETURN_TYPES
+from nncf.torch.return_types import maybe_unwrap_from_torch_return_type
+from nncf.torch.return_types import maybe_wrap_to_torch_return_type
+
+
+def test_unwrap_wrap_torch_return_type():
+    for return_type in _TORCH_RETURN_TYPES:
```
Comment (AlexanderDokuchaev) [inline tests/torch/test_return_types.py:19]: Use pytest parameters for return_type

File: nncf/torch/return_types.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+def maybe_unwrap_from_torch_return_type(tensor: Any) -> torch.Tensor:
+    """
+    Attempts to unwrap the tensor value from one of torch.return_types instantces
+    in case torch operation output is wrapped by a torch return_type.
+
+    :param tensor: Torch tensor or torch return type instance to unwrap values from.
+    :return: Unwrapped torch tensor.
+    """
+    if isinstance(tensor, _TORCH_RETURN_TYPES):
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/return_types.py:40]: I believe that the check `isinstance(tensor, _TORCH_RETURN_TYPES)` must be present as check `hasattr(tensor, 'value')` could lead to errors. Besides `_TORCH_RETURN_TYPES`, there are bunch of classes with `value` attribute (ex: dict).

File: tests/torch/test_return_types.py (hunk 3 above)
Comment (daniil-lyakhov) [reply tests/torch/test_return_types.py]: Done

File: nncf/torch/return_types.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Tuple, Type, Union
+
+import torch
+
+
+def __get_supported_torch_return_types() -> Tuple[Type[object]]:
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/return_types.py]: Done

File: nncf/torch/return_types.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    """
+    Collects types from torch.return_type which can be wrapped/unwrapped by nncf.
+
+    :return: List of types from torch.return_type which can be wrapped/unwrapped by nncf.
+    """
+    return_type_names = [t for t in dir(torch.return_types) if not t.startswith("_") and not t.startswith("linalg")]
+    return_types = [getattr(torch.return_types, t_name) for t_name in return_type_names]
+    return_types = [t for t in return_types if hasattr(t, "values")]
+    return tuple(return_types)
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/torch/return_types.py:27]: Done

**Comment by daniil-lyakhov** [comment]
test compression_training build 129 is finished successfully

**Review by AlexanderDokuchaev — APPROVED** [review]

File: nncf/torch/return_types.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    Wraps tensor to wrapped_input wrapper in case wrapped_input is wrapped by a torch.return_value container.
+
+    :param tensor: Torch tensor to wrap.
+    :param wrapped_tensor: Instance of the tensor before it was unwrapped.
+    :return: Wrapped tensor in case wrapped_input is wrapped by a torch.return_value container else the tensor.
+    """
+
+    if isinstance(wrapped_input, _TORCH_RETURN_TYPES):
+        return wrapped_input.__class__([tensor, *wrapped_input[1:]])
```
Comment (ljaljushkin) [inline nncf/torch/return_types.py:54]: Why does it always start from first axis?

Comment (ljaljushkin) [reply nncf/torch/return_types.py:54]: Agreed to elaborate tricky part with values in the comment and use 2 inputs only.

**Review by ljaljushkin — APPROVED** [review]

File: nncf/torch/return_types.py (hunk 7 above)
Comment (daniil-lyakhov) [reply nncf/torch/return_types.py]: Thanks, done
